### ISSUE TITLE
ArrayIndexOutOfBoundsException in LCSMatchingAlgo - Kotlin Multiplatform

### ISSUE THREAD

**Issue by mdivkovic-speech** [mockk/mockk#600]
### Expected Behavior

A more meaningful error message, to be able to fix my test.

### Current Behavior

Once in around every 10 test runs in our CI pipeline in GitHub Actions, tests fail with this exception. Not always the same test, but one out of a couple tests. I haven't been able to detect a pattern or commonality between them, beyond the obvious fact that they all involve coroutines.

### Failure Information (for bugs)

As stated in the title, this is part of a Kotlin Multiplatform Project, for whatever that's worth.
The fact that it doesn't happen every time, leads me to believe that there is a race condition somewhere in my test code, but I would really appreciate a more precise message from MockK instead of it jus crashing.

#### Steps to Reproduce

I wish I knew...

#### Context

* MockK version: 1.11.0
* OS: Ubuntu-20.04
* Kotlin version: 1.4.32
* JDK version: 1.8
* JUnit version: 4.13.2
* Type of test: Android Unit Test

#### Stack trace

```
2021-04-13T17:07:27.5689742Z login.LoginPresenterTest > before attempting login tells view to show loading indicator FAILED
2021-04-13T17:07:27.5692172Z     java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 0
2021-04-13T17:07:27.5694120Z         at io.mockk.impl.verify.LCSMatchingAlgo.lcs(LCSMatchingAlgo.kt:52)
2021-04-13T17:07:27.5713535Z         at io.mockk.impl.verify.VerificationHelpers.reportCalls$default(VerificationHelpers.kt:75)
2021-04-13T17:07:27.5716463Z         at io.mockk.impl.verify.OrderedCallVerifier$verify$1.invoke(OrderedCallVerifier.kt:28)
2021-04-13T17:07:27.5718605Z         at io.mockk.impl.verify.OrderedCallVerifier$verify$1.invoke(OrderedCallVerifier.kt:12)
2021-04-13T17:07:27.5720944Z         at io.mockk.impl.recording.CommonCallRecorder.safeExec(CommonCallRecorder.kt:72)
2021-04-13T17:07:27.5722998Z         at io.mockk.impl.log.SafeToString.exec(SafeToString.kt:10)
2021-04-13T17:07:27.5724967Z         at io.mockk.impl.verify.OrderedCallVerifier.verify(OrderedCallVerifier.kt:26)
2021-04-13T17:07:27.5733629Z         at io.mockk.impl.verify.TimeoutVerifier.verify(TimeoutVerifier.kt:22)
2021-04-13T17:07:27.5735916Z         at io.mockk.impl.recording.states.VerifyingState$recordingDone$outcome$1.invoke(VerifyingState.kt:31)
2021-04-13T17:07:27.5738271Z         at io.mockk.impl.recording.states.VerifyingState$recordingDone$outcome$1.invoke(VerifyingState.kt:11)
2021-04-13T17:07:27.5740684Z         at io.mockk.impl.recording.CommonCallRecorder.safeExec(CommonCallRecorder.kt:72)
2021-04-13T17:07:27.5743311Z         at io.mockk.impl.recording.states.VerifyingState.recordingDone(VerifyingState.kt:30)
2021-04-13T17:07:27.5745828Z         at io.mockk.impl.recording.CommonCallRecorder.done(CommonCallRecorder.kt:47)
2021-04-13T17:07:27.5748461Z         at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:64)
2021-04-13T17:07:27.5750797Z         at io.mockk.impl.eval.VerifyBlockEvaluator.verify(VerifyBlockEvaluator.kt:30)
2021-04-13T17:07:27.5753464Z         at io.mockk.MockKDsl.internalCoVerify(API.kt:143)
2021-04-13T17:07:27.5754659Z         at io.mockk.MockKKt.coVerify(MockK.kt:175)
2021-04-13T17:07:27.5755734Z         at io.mockk.MockKKt.coVerify$default(MockK.kt:172)
2021-04-13T17:07:27.5758972Z         at login.LoginPresenterTest.before attempting login tells view to show loading indicator(LoginPresenterTest.kt:166)
```

#### Minimal reproducible code (the gist of this issue)

As I said above, the error is not reliably reproducible, but the test case that causes it, and the SUT code that's being tested looks like this:

```
// -----------------------[ GRADLE DEFINITIONS ] -----------------------
        commonMain {
            dependencies {
                api("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.4.3-native-mt")
            }
        }
        commonTest {
            dependencies {
                implementation(kotlin("test-common"))
                implementation(kotlin("test-annotations-common"))
                implementation("io.mockk:mockk-common:1.11.0")
            }
        }

        sourceSets["androidTest"].dependencies {
            implementation(kotlin("test"))
            implementation(kotlin("test-junit"))
            implementation("junit:junit:4.13.2")
            implementation("io.mockk:mockk:1.11.0")
        }
// -----------------------[ YOUR CODE STARTS HERE ] -----------------------

// Simplified test class looks like this:

package login

import io.mockk.*
import login.usecase.LogInUser
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.asCoroutineDispatcher
import java.util.concurrent.Executors

class LoginPresenterTest {

private val testDispatcher : CoroutineDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()

// LoginView has all Unit functions
private lateinit var view: LoginView

private lateinit var logInUser: LogInUser

@BeforeTest
fun setup() {
        logInUser = mockk()
        view = mockk(relaxUnitFun = true)
}

// Failing test case
fun `tells view to show loading indicator before attempting login`() {
        val presenter = LoginPresenter(testDispatcher, logInUser)
        coEvery { logInUser(any(), any()) } returns LogInUser.Output.Success("User ID")
        presenter.attachView(view)
        presenter.onLogInClick("email", "password")
        coVerify(timeout = 500, ordering = Ordering.ORDERED) {
            view.showLoadingIndicator()
            logInUser("email", "password")
        }
    }
}


// Simplified presenter looks like this:

package login

import kotlinx.coroutines.*
import util.NonCapturingCoroutineExceptionHandler
import util.log.Logger
import kotlin.coroutines.CoroutineContext
import kotlinx.coroutines.CoroutineExceptionHandler
import login.usecase.LogInUser
import login.LoginView

class LoginPresenter (
    private val mainContext: CoroutineContext,
    private val logInUser: LogInUser
) :  CoroutineScope {

    var view: LoginView? = null
    private val job = SupervisorJob()
    private val exceptionHandler = CoroutineExceptionHandler { _, throwable ->
        throwable.printStackTrace()
    }

    fun attachView(view: LoginView) {
        this.view = view
    }

    fun detachView() {
        view = null
    }

    fun cancelChildJobs() {
        job.cancelChildren()
    }

    override fun onFocus() {
        logger.i("Focused")
    }

    override fun onFocusLost() {
        logger.i("Focus lost")
    }

    override val coroutineContext: CoroutineContext
        get() = mainContext + job + exceptionHandler.exceptionHandler

    fun onLogInClick(email: String, password: String) {
          launch {
              view?.showLoadingIndicator()
              val output = logInUser(email, password)
              view?.hideLoadingIndicator()
              when (output) {
                  is LogInUser.Output.Success -> view?.showList()
                  LogInUser.Output.UnexpectedError -> view?.showUnexpectedError()
              }
         }
    }
}

// Simplified LoginView interface looks like this:

package login

interface LoginView {

    fun showLoadingIndicator()

    fun hideLoadingIndicator()

    fun showUnexpectedError()

    fun showList()
}

// Simplified LogInUser use case looks like this:

package login.usecase

import login.entity.User

interface LogInUser {

    suspend operator fun invoke(email: String, password: String) : Output

    sealed class Output {
        data class Success(val userId: String) : Output()
        object UnexpectedError: Output()
    }
}

// -----------------------[ YOUR CODE ENDS HERE ] -----------------------
```

**Comment by Raibaz** [comment]
I'm guessing there's some kind of conflict between different tests in your test class.

As a random test, can you please try calling `unmockkAll()` in an @AfterTest function?

**Comment by mdivkovic-speech** [comment]
Thank you for the suggestion, I will try it fiest thing tomorrow morning.

**Comment by mdivkovic-speech** [comment]
@Raibaz I've added a teardown method with the @AfterTest annotation and I call unmockAll and clearAllMocks in it. I will have to run several iterations of the tests to see if it makes a difference.

**Comment by mdivkovic-speech** [comment]
OK, I got the same failure now despite the new @AfterTest. 

To be fair, I added this change only to the two classes where the tests were failing, to try it out quickly. Do we have to do it for all of them? Maybe the other 500+ test cases which are creating mocks are putting too much load on MockK and it breaks in these tests specifically. Do you have any experience with that?

**Comment by Raibaz** [comment]
I don't think different test classes may be affecting each other, unless you are sharing variables between them.

As a temporary workaround, can you please try removing `ordering = Ordering.ORDERED` from your verify block?

Ordered verification uses a different matching algorithm and I'd like to see if the problem is specific to this algorithm or if it's unrelated.

Thanks!

**Comment by mdivkovic-speech** [comment]
@Raibaz I think I found the cause and it has nothing to do with mockk. In our tests with coroutines we were using a single thread coroutine dispatcher to make our tests execute predictably, BUT not the official one that comes with the kotlinx-coroutines library. Turns out that official one has some other magic going on under the hood, and after switching to it, we cannot reproduce this issue anymore.

I'm closing it. Thank you very much for your time and suggestions!
